apis/placement/v1alpha1/stagedupdate_types.go

```diff
@@ -0,0 +1,478 @@
+
+	// The time when the update started on the stage. Empty if the stage has not started updating.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Type=string
+	// +kubebuilder:validation:Format=date-time
+	StartTime *metav1.Time `json:"startTime,omitempty"`
+
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	//
+	// Conditions is an array of current observed updating conditions for the stage. Empty if the stage has not started updating.
+	// Known conditions are "Progressing", "Succeeded".
+	// +kubebuilder:validation:Optional
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+
+// StageUpdatingConditionType identifies a specific condition of the stage that is being updated.
+// +enum
+type StageUpdatingConditionType string
+
+const (
+	// StageUpdatingConditionProgressing indicates whether the stage updating is making progress.
+	// Its condition status can be one of the following:
+	// - "True" means the stage updating is making progress.
+	// - "False" means the stage updating is waiting/pausing.
+	StageUpdatingConditionProgressing StageUpdatingConditionType = "Progressing"
+
+	// ClusterUpdatingStatusConditionSucceeded indicates whether the stage updating is completed successfully.
+	// Its condition status can be one of the following:
+	// - "True" means the stage updating is completed successfully.
+	// - "False" means the stage updating encountered an error and stopped.
+	ClusterUpdatingStatusConditionSucceeded StageUpdatingConditionType = "Succeeded"
+)
+
+// ClusterUpdatingStatus defines the status of the update run on a cluster.
+type ClusterUpdatingStatus struct {
+	// The name of the cluster.
+	// +kubebuilder:validation:Required
+	ClusterName string `json:"clusterName"`
+
+	// ResourceOverrideSnapshots is a list of ResourceOverride snapshots associated with the cluster.
+	// The list is computed at the beginning of the update run and not updated during the update run.
+	// The list is empty if there are no resource overrides associated with the cluster.
+	// +kubebuilder:validation:Optional
+	ResourceOverrideSnapshots []v1beta1.NamespacedName `json:"resourceOverrideSnapshots,omitempty"`
+
+	// ClusterResourceOverrides contains a list of applicable ClusterResourceOverride snapshot names
+	// associated with the cluster.
+	// The list is computed at the beginning of the update run and not updated during the update run.
+	// The list is empty if there are no cluster overrides associated with the cluster.
+	// +kubebuilder:validation:Optional
+	ClusterResourceOverrideSnapshots []string `json:"clusterResourceOverrideSnapshots,omitempty"`
+
+	// The time when the update started on the cluster. Empty if the cluster has not started updating.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Type=string
+	// +kubebuilder:validation:Format=date-time
+	StartTime *metav1.Time `json:"startTime,omitempty"`
```

Comment:
Can we add endtime?

Reply:
the end time is the lastTransitionTime of the "succeded" condition